Help suggests non-existent package std::os::ext::process::CommandExt

Normally rustc will suggest what `use` line to add, but for CommandExt its suggestion is impossible.

testcase.rs:
```rust
use std::process::Command;
// forgot this: use std::os::unix::process::CommandExt;

fn main() {
    Command::new("echo").arg("hello").exec();
}
```
Compile with `rustc testcase.rs`.  Rust says it can't find `exec` and explains:

> ```
>   = help: items from traits can only be used if the trait is in scope; the following trait is implemented but not in scope, perhaps add a `use` for it:
>   = help: candidate #1: `use std::os::ext::process::CommandExt;`
> ```

But the suggested trait doesn't exist; adding the suggested line gives an unresolved import error: "Could not find `ext` in `os`"

Rust should instead suggest the correct trait, `std::os::unix::process::CommandExt`.

Tested with (`rustc --version --verbose`):

Fedora 25's packaged rust:
> ```
> rustc 1.10.0 (cfcb716cf 2016-07-03)
> binary: rustc
> commit-hash: cfcb716cf0961a7e3a4eceac828d94805cf8140b
> commit-date: 2016-07-03
> host: x86_64-unknown-linux-gnu
> release: 1.10.0
> ```

And a recent source build:

> ```
> rustc 1.16.0-dev (74c42ac17 2017-01-19)
> binary: rustc
> commit-hash: 74c42ac173bee900979870ed986c760596d1fbdb
> commit-date: 2017-01-19
> host: x86_64-unknown-linux-gnu
> release: 1.16.0-dev
> LLVM version: 3.9
> ```

(The formatting of the error messages differs between versions but the errors are the same.)

Other modules show this problem:

```rust
error[E0432]: unresolved import `std::os::ext::ffi::OsStrExt`
  --> src/main.rs:44:5
   |
44 | use std::os::ext::ffi::OsStrExt;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^ Could not find `ext` in `os`

error: no method named `as_bytes` found for type `&std::ffi::OsStr` in the current scope
  --> src/main.rs:51:41
   |
51 | 		.filter(|n| *(n.unwrap().file_name()).as_bytes() == b".dsc") {
   | 		                                      ^^^^^^^^
   |
   = help: items from traits can only be used if the trait is in scope; the following trait is implemented but not in scope, perhaps add a `use` for it:
   = help: candidate #1: `use std::os::ext::ffi::OsStrExt;`

error: aborting due to previous error
```

Thanks

Just ran into this with `1.28.0` on windows 10. I guess I'll add my example to the pile:
```rust
use std::ffi::OsStr;
use std::os::ext::ffi::OsStrExt;

fn main() {
    OsStr::new("1234").encode_wide();
}
```

This gives me the following absurd pair of errors:
```
error[E0433]: failed to resolve. Could not find `ext` in `os`
 --> src\main.rs:2:14
  |
2 | use std::os::ext::ffi::OsStrExt;
  |              ^^^ Could not find `ext` in `os`

warning: unused import: `std::os::ext::ffi::OsStrExt`
 --> src\main.rs:2:5
  |
2 | use std::os::ext::ffi::OsStrExt;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: #[warn(unused_imports)] on by default

error[E0599]: no method named `encode_wide` found for type `&std::ffi::OsStr` in the current scope
 --> src\main.rs:5:24
  |
5 |     OsStr::new("1234").encode_wide();
  |                        ^^^^^^^^^^^
  |
  = help: items from traits can only be used if the trait is in scope
help: the following trait is implemented but not in scope, perhaps add a `use` for it:
  |
1 | use std::os::ext::ffi::OsStrExt;
  |

error: aborting due to 2 previous errors
```

I likewise get this issue with the following snippet:

```rust
use std::ffi::OsString;

fn main() {
    let _ = OsString::from_wide(b"a\x00b\x00c\x00");
}
```

```
   Compiling bad_ffi_ext_trait_suggestion v0.1.0 (file:///C:/Users/egubler/AppData/Local/Cargo/script-cache/file-bad_ffi_ext_trait_suggestion-eaa1f7e82798a3de)
error[E0599]: no function or associated item named `from_wide` found for type `std::ffi::OsString` in the current scope
 --> bad_ffi_ext_trait_suggestion.rs:4:13
  |
4 |     let _ = OsString::from_wide(b"a\x00b\x00c\x00");
  |             ^^^^^^^^^^^^^^^^^^^ function or associated item not found in `std::ffi::OsString`
  |
  = help: items from traits can only be used if the trait is in scope
help: the following trait is implemented but not in scope, perhaps add a `use` for it:
  |
1 | use std::os::ext::ffi::OsStringExt;
  |

error: aborting due to previous error

For more information about this error, try `rustc --explain E0599`.
error: Could not compile `bad_ffi_ext_trait_suggestion`.

To learn more, run the command again with --verbose.
```

--- a/src/librustc/ty/item_path.rs
+++ b/src/librustc/ty/item_path.rs
@@ -10,7 +10,7 @@
 
 use hir::map::DefPathData;
 use hir::def_id::{CrateNum, DefId, CRATE_DEF_INDEX, LOCAL_CRATE};
-use ty::{self, Ty, TyCtxt};
+use ty::{self, DefIdTree, Ty, TyCtxt};
 use middle::cstore::{ExternCrate, ExternCrateSource};
 use syntax::ast;
 use syntax::symbol::{keywords, LocalInternedString, Symbol};
@@ -219,19 +219,73 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
                 cur_def_key = self.def_key(parent);
             }
 
+            let visible_parent = visible_parent_map.get(&cur_def).cloned();
+            let actual_parent = self.parent(cur_def);
+            debug!(
+                "try_push_visible_item_path: visible_parent={:?} actual_parent={:?}",
+                visible_parent, actual_parent,
+            );
+
             let data = cur_def_key.disambiguated_data.data;
-            let symbol = data.get_opt_name().map(|n| n.as_str()).unwrap_or_else(|| {
-                if let DefPathData::CrateRoot = data { // reexported `extern crate` (#43189)
-                    self.original_crate_name(cur_def.krate).as_str()
-                } else {
-                    Symbol::intern("<unnamed>").as_str()
-                }
-            });
+            let symbol = match data {
+                // In order to output a path that could actually be imported (valid and visible),
+                // we need to handle re-exports correctly.
+                //
+                // For example, take `std::os::unix::process::CommandExt`, this trait is actually
+                // defined at `std::sys::unix::ext::process::CommandExt` (at time of writing).
+                //
+                // `std::os::unix` rexports the contents of `std::sys::unix::ext`. `std::sys` is
+                // private so the "true" path to `CommandExt` isn't accessible.
+                //
+                // In this case, the `visible_parent_map` will look something like this:
+                //
+                // (child) -> (parent)
+                // `std::sys::unix::ext::process::CommandExt` -> `std::sys::unix::ext::process`
+                // `std::sys::unix::ext::process` -> `std::sys::unix::ext`
+                // `std::sys::unix::ext` -> `std::os`
+                //
+                // This is correct, as the visible parent of `std::sys::unix::ext` is in fact
+                // `std::os`.
+                //
+                // When printing the path to `CommandExt` and looking at the `cur_def_key` that
+                // corresponds to `std::sys::unix::ext`, we would normally print `ext` and then go
+                // to the parent - resulting in a mangled path like
+                // `std::os::ext::process::CommandExt`.
+                //
+                // Instead, we must detect that there was a re-export and instead print `unix`
+                // (which is the name `std::sys::unix::ext` was re-exported as in `std::os`). To
+                // do this, we compare the parent of `std::sys::unix::ext` (`std::sys::unix`) with
+                // the visible parent (`std::os`). If these do not match, then we iterate over
+                // the children of the visible parent (as was done when computing
+                // `visible_parent_map`), looking for the specific child we currently have and then
+                // have access to the re-exported name.
+                DefPathData::Module(module_name) if visible_parent != actual_parent => {
+                    let mut name: Option<ast::Ident> = None;
+                    if let Some(visible_parent) = visible_parent {
+                        for child in self.item_children(visible_parent).iter() {
+                            if child.def.def_id() == cur_def {
+                                name = Some(child.ident);
+                            }
+                        }
+                    }
+                    name.map(|n| n.as_str()).unwrap_or(module_name.as_str())
+                },
+                _ => {
+                    data.get_opt_name().map(|n| n.as_str()).unwrap_or_else(|| {
+                        // Re-exported `extern crate` (#43189).
+                        if let DefPathData::CrateRoot = data {
+                            self.original_crate_name(cur_def.krate).as_str()
+                        } else {
+                            Symbol::intern("<unnamed>").as_str()
+                        }
+                    })
+                },
+            };
             debug!("try_push_visible_item_path: symbol={:?}", symbol);
             cur_path.push(symbol);
 
-            match visible_parent_map.get(&cur_def) {
-                Some(&def) => cur_def = def,
+            match visible_parent {
+                Some(def) => cur_def = def,
                 None => return false,
             };
         }

--- a/src/test/ui/issues/issue-39175.rs
+++ b/src/test/ui/issues/issue-39175.rs
@@ -1,0 +1,25 @@
+// Copyright 2014 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// This test ignores some platforms as the particular extension trait used
+// to demonstrate the issue is only available on unix. This is fine as
+// the fix to suggested paths is not platform-dependent and will apply on
+// these platforms also.
+
+// ignore-windows
+// ignore-cloudabi
+// ignore-emscripten
+
+use std::process::Command;
+// use std::os::unix::process::CommandExt;
+
+fn main() {
+    Command::new("echo").arg("hello").exec();
+}

--- a/src/test/ui/issues/issue-39175.stderr
+++ b/src/test/ui/issues/issue-39175.stderr
@@ -1,0 +1,15 @@
+error[E0599]: no method named `exec` found for type `&mut std::process::Command` in the current scope
+  --> $DIR/issue-39175.rs:24:39
+   |
+LL |     Command::new("echo").arg("hello").exec();
+   |                                       ^^^^
+   |
+   = help: items from traits can only be used if the trait is in scope
+help: the following trait is implemented but not in scope, perhaps add a `use` for it:
+   |
+LL | use std::os::unix::process::CommandExt;
+   |
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0599`.

